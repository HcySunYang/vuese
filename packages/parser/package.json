{
  "name": "@vuese/parser",
  "version": "2.6.0",
  "description": "Vue file parser for automatic document generation",
  "keywords": [
    "parser",
    "vue",
    "document"
  ],
  "author": "HcySunYang <HcySunYang@outlook.com>",
  "homepage": "https://github.com/vuese/vuese/tree/master/packages/parser#readme",
  "license": "MIT",
  "main": "dist/lib/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "tsc": "rm -rf dist && tsc"
  },
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuese/vuese.git"
  },
  "bugs": {
    "url": "https://github.com/vuese/vuese/issues"
  },
  "buildOptions": {
    "extractDts": true
  },
  "dependencies": {
    "@babel/generator": "~7.4.4",
    "@babel/parser": "~7.4.4",
    "@babel/traverse": "~7.4.4",
    "@babel/types": "~7.4.4",
    "vue-template-compiler": "^2.6.10"
  },
  "gitHead": "2bc1b0691aaf9480fa6fc0ad59ff1bc991b81efe"
}
