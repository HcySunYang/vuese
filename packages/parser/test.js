const a = require('../parser/dist/index.js')

console.log(a.parser('test'))
