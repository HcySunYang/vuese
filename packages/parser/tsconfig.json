{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": ".",
    "module": "CommonJS",
    "outDir": "dist",
    "moduleResolution": "node",
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "target": "es2017",
    "esModuleInterop": true,
    "incremental": false
  },
  "include": ["lib", "../declarations.d.ts", "__test__"],
}
