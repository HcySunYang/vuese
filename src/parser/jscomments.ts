import { isCommentLine, isCommentBlock } from '../helpers'
import * as bt from '@babel/types'

export type CommentResult = {
  default: string[]
  [key: string]: string[]
}
const commentRE = /\n\s*\*?\s*/g
const leadRE = /^@(\w+)\b/

export function getComments(cnode: bt.Node): CommentResult {
  const res: CommentResult = {
    default: []
  }
  const commentNodes = cnode.leadingComments || []
  if (!commentNodes || !commentNodes.length) return res

  let comments: string = '',
    matchs: RegExpMatchArray | null
  ;(commentNodes as []).forEach((node: bt.Comment) => {
    if (isCommentLine(node)) {
      comments = node.value.trim()
      matchs = comments.match(leadRE)
      if (matchs) {
        const key: string = matchs[1]
        res[key] = res[key] || []
        res[key].push(comments.replace(leadRE, ''))
      } else {
        res.default.push(comments)
      }
    } else if (isCommentBlock(node)) {
      comments = node.value
        .replace(commentRE, '\n')
        .replace(/^\*/, '')
        .trim()
      let currentKey = 'default'
      comments.split('\n').forEach(c => {
        if ((matchs = c.match(leadRE))) {
          currentKey = matchs[1]
          res[currentKey] = res[currentKey] || []
          res[currentKey].push(c.replace(leadRE, '').trim())
        } else {
          res.default.push(c)
        }
      })
    }
  })
  return res
}
