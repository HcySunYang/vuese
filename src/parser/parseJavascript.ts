import traverse, { NodePath } from '@babel/traverse'
import generate from '@babel/generator'
import * as bt from '@babel/types'
import { getComments, CommentResult } from './jscomments'
import {
  PropsResult,
  PropType,
  ParserOptions,
  EventResult,
  MethodResult
} from './index'
import {
  getValueFromGenerate,
  isVueOption,
  runFunction,
  getFirstPropContainer
} from '../helpers'
import { isArray } from 'util'

function normalizeProps(props: string[]): PropsResult[] {
  return props.map(prop => ({
    type: null,
    name: prop
  }))
}

function getTypeByPath(typePath: NodePath): PropType {
  const typeNode = typePath.node
  if (bt.isIdentifier(typeNode)) return typeNode.name
  if (bt.isArrayExpression(typeNode)) {
    if (!typeNode.elements.length) return null

    const container = getFirstPropContainer(typePath, 'elements')

    return container
      .filter(node => bt.isIdentifier(node))
      .map(node => (node as bt.Identifier).name)
  }

  return null
}

// The `type` of a prop should be an array of constructors or constructors
// eg. String or [String, Number]
function isAllowPropsType(typeNode: bt.Node): boolean {
  return bt.isIdentifier(typeNode) || bt.isArrayExpression(typeNode)
}

function hasFunctionTypeDef(type: PropType): boolean {
  if (typeof type === 'string') {
    return type.toLowerCase() === 'function'
  } else if (isArray(type)) {
    return type.map(a => a.toLowerCase()).some(b => b === 'function')
  }
  return false
}

export default function(ast: bt.File, options: ParserOptions = {}) {
  const seenEvent = new Set()

  traverse(ast, {
    ExportDefaultDeclaration(path: NodePath<bt.ExportDefaultDeclaration>) {
      path.traverse(
        {
          ObjectProperty(path: NodePath<bt.ObjectProperty>) {
            const { onProp, onMethod, onName } = this.options
            // Processing name
            if (isVueOption(path, 'name')) {
              if (onName) onName((path.node.value as bt.StringLiteral).value)
            }

            // Processing props
            if (isVueOption(path, 'props')) {
              const valuePath = path.get('value')

              let res: PropsResult[] = []
              if (bt.isArrayExpression(valuePath.node)) {
                // An array of strings
                const propsValue: [] = getValueFromGenerate(valuePath.node)
                res = normalizeProps(propsValue)
              } else if (bt.isObjectExpression(valuePath.node)) {
                // An object
                valuePath.traverse({
                  ObjectProperty(propPath: NodePath<bt.ObjectProperty>) {
                    // Guarantee that this is the prop definition
                    if (propPath.parentPath === valuePath) {
                      const name = propPath.node.key.name
                      const vPath = propPath.get('value')
                      const result: PropsResult = {
                        name,
                        type: null,
                        describe: getComments(propPath.node).default
                      }

                      if (isAllowPropsType(vPath.node)) {
                        result.type = getTypeByPath(vPath)
                      } else if (bt.isObjectExpression(vPath.node)) {
                        if (!vPath.node.properties.length) return

                        const allPropNodes = getFirstPropContainer(
                          vPath,
                          'properties'
                        )

                        const typeNode: any[] = allPropNodes.filter(
                          (node: any, i: number) => {
                            if (node.key.name === 'type') {
                              node.$$selfPath = vPath.get(`properties.${i}`)
                              return true
                            }
                            return false
                          }
                        )
                        const otherNodes = allPropNodes.filter(
                          (node: any, i: number) => {
                            if (node.key.name !== 'type') {
                              node.$$selfPath = vPath.get(`properties.${i}`)
                              return true
                            }
                            return false
                          }
                        )

                        // Prioritize `type` before processing `default`.
                        // Because the difference in `type` will affect the way `default` is handled.
                        if (typeNode.length > 0) {
                          result.type = getTypeByPath(
                            typeNode[0].$$selfPath.get('value')
                          )
                          // Get descriptions of the type
                          const typeDesc: string[] = getComments(typeNode[0])
                            .default
                          if (typeDesc.length > 0) {
                            result.typeDesc = typeDesc
                          }
                        }

                        otherNodes.forEach((node: any) => {
                          const n = node.key.name
                          if (n === 'default') {
                            if (
                              !hasFunctionTypeDef(result.type) &&
                              bt.isFunction(node.value)
                            ) {
                              result.default = runFunction(node.value)
                            } else {
                              if (bt.isObjectMethod(node)) {
                                result.default = generate(node).code
                              } else {
                                result.default = generate(node.value).code
                              }
                            }

                            // Get descriptions of the default value
                            const defaultDesc: string[] = getComments(node)
                              .default
                            if (defaultDesc.length > 0) {
                              result.defaultDesc = defaultDesc
                            }
                          } else if (n === 'required') {
                            if (bt.isBooleanLiteral(node.value)) {
                              result.required = node.value.value
                            }
                          } else if (n === 'validator') {
                            if (bt.isObjectMethod(node)) {
                              result.validator = generate(node).code
                            } else {
                              result.validator = generate(node.value).code
                            }

                            // Get descriptions of the validator
                            const validatorDesc: string[] = getComments(node)
                              .default
                            if (validatorDesc.length > 0) {
                              result.validatorDesc = validatorDesc
                            }
                          }
                        })
                      }
                      res.push(result)
                    }
                  }
                })
              }
              if (onProp) onProp(res)
            }

            // Processing methods
            if (isVueOption(path, 'methods')) {
              const properties = (path.node
                .value as bt.ObjectExpression).properties.filter(
                n => bt.isObjectMethod(n) || bt.isObjectProperty(n)
              ) as (bt.ObjectMethod | bt.ObjectProperty)[]

              properties.forEach(node => {
                const commentsRes: CommentResult = getComments(node)
                // Collect only methods that have @vuese annotations
                if (commentsRes.vuese) {
                  const result: MethodResult = {
                    name: node.key.name,
                    describe: commentsRes.default,
                    argumentsDesc: commentsRes.arg
                  }
                  if (onMethod) onMethod(result)
                }
              })
            }
          },
          CallExpression(path: NodePath<bt.CallExpression>) {
            const node = path.node
            // this.$emit()
            if (
              bt.isMemberExpression(node.callee) &&
              bt.isThisExpression(node.callee.object) &&
              bt.isIdentifier(node.callee.property) &&
              node.callee.property.name === '$emit' &&
              bt.isExpressionStatement(path.parentPath.node)
            ) {
              const { onEvent } = this.options
              const args = node.arguments
              const result: EventResult = {
                name: ''
              }
              const firstArg = args[0]

              if (args.length && bt.isStringLiteral(firstArg)) {
                result.name = firstArg.value
              }

              if (!result.name || this.seenEvent.has(result.name)) return
              this.seenEvent.add(result.name)

              const allComments: CommentResult = getComments(
                path.parentPath.node
              )
              result.describe = allComments.default
              result.argumentsDesc = allComments.arg

              if (onEvent) onEvent(result)
            }
          }
        },
        {
          options,
          seenEvent
        }
      )
    }
  })
}
