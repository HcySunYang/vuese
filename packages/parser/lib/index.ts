import { sfcToAST } from './sfcToAST'
import { parseJavascript } from './parseJavascript'
import { parseTemplate } from './parseTemplate'
import { ParserPlugin } from '@babel/parser'
import { CommentResult } from './jscomments'

export * from './sfcToAST'
export * from './parseJavascript'
export * from './parseTemplate'
export * from './helper'
export * from './jscomments'

export type PropType = string | string[] | null

export type BabelParserPlugins = { [key in ParserPlugin]?: boolean }

export interface PropsResult {
  type: PropType
  name: string
  typeDesc?: string[]
  required?: boolean
  default?: string
  defaultDesc?: string[]
  validator?: string
  validatorDesc?: string[]
  describe?: string[]
}

export interface EventResult {
  name: string
  isSync: boolean
  syncProp: string
  describe?: string[]
  argumentsDesc?: string[]
}

export interface MethodResult {
  name: string
  describe?: string[]
  argumentsDesc?: string[]
}

export interface ComputedResult {
  name: string
  describe?: string[]
}

export interface MixInResult {
  mixIn: string
}

export type AttrsMap = {
  [key: string]: string
}

export interface SlotResult {
  name: string
  describe: string
  backerDesc: string
  bindings: AttrsMap
  scoped: boolean
}

export interface ParserOptions {
  onProp?: {
    (propsRes: PropsResult): void
  }
  onEvent?: {
    (eventRes: EventResult): void
  }
  onMethod?: {
    (methodRes: MethodResult): void
  }
  onComputed?: {
    (computedRes: ComputedResult): void
  }
  onMixIn?: {
    (mixInRes: MixInResult): void
  }
  onSlot?: {
    (slotRes: SlotResult): void
  }
  onName?: {
    (name: string): void
  }
  onDesc?: {
    (desc: CommentResult): void
  }
  babelParserPlugins?: BabelParserPlugins
}

export interface ParserResult {
  props?: PropsResult[]
  events?: EventResult[]
  slots?: SlotResult[]
  mixIns?: MixInResult[]
  methods?: MethodResult[]
  computed?: ComputedResult[]
  name?: string
  componentDesc?: CommentResult
}

export function parser(
  source: string,
  options: ParserOptions = {}
): ParserResult {
  const astRes = sfcToAST(source, options.babelParserPlugins)
  const res: ParserResult = {}
  const defaultOptions: ParserOptions = {
    onName(name: string) {
      res.name = name
    },
    onDesc(desc: CommentResult) {
      res.componentDesc = desc
    },
    onProp(propsRes: PropsResult) {
      ;(res.props || (res.props = [])).push(propsRes)
    },
    onEvent(eventsRes: EventResult) {
      ;(res.events || (res.events = [])).push(eventsRes)
    },
    onSlot(slotRes: SlotResult) {
      ;(res.slots || (res.slots = [])).push(slotRes)
    },
    onMixIn(mixInRes: MixInResult) {
      ;(res.mixIns || (res.mixIns = [])).push(mixInRes)
    },
    onMethod(methodRes: MethodResult) {
      ;(res.methods || (res.methods = [])).push(methodRes)
    },
    onComputed(computedRes: ComputedResult) {
      ;(res.computed || (res.computed = [])).push(computedRes)
    }
  }

  const finallyOptions: ParserOptions = Object.assign(defaultOptions, options)
  if (astRes.jsAst) {
    parseJavascript(astRes.jsAst, finallyOptions)
  }
  if (astRes.templateAst) {
    parseTemplate(astRes.templateAst, finallyOptions)
  }

  return res
}
