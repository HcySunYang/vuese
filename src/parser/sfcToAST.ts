import { parse as vueSFCParse } from '@vue/component-compiler-utils'
import * as vueTemplateCompiler from 'vue-template-compiler'
import { parse as babelParse, ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'

export interface AstResult {
  jsAst?: bt.File
  templateAst?: object
}

export default function(
  source: string,
  babelParserPlugins?: ParserPlugin[]
): AstResult {
  const sfc = vueSFCParse({
    source: source,
    compiler: vueTemplateCompiler,
    needMap: false
  })
  const res: AstResult = {}
  if (sfc.script && sfc.script.content) {
    res.jsAst = babelParse(sfc.script.content, {
      sourceType: 'module',
      plugins: babelParserPlugins || ['objectRestSpread', 'dynamicImport']
    })
  }
  if (sfc.template && sfc.template.content) {
    res.templateAst = vueTemplateCompiler.compile(sfc.template.content, {
      comments: true
    }).ast
  }
  return res
}
