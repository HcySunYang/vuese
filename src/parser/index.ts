import sfcToAST from './sfcToAST'
import parseJavascript from './parseJavascript'
import parseTemplate from './parseTemplate'
import { ParserPlugin } from '@babel/parser'

export type PropType = string | string[] | null

export interface PropsResult {
  type: PropType
  name: string
  typeDesc?: string[]
  required?: boolean
  default?: string
  defaultDesc?: string[]
  validator?: string
  validatorDesc?: string[]
  describe?: string[]
}

export interface EventResult {
  name: string
  describe?: string[]
  argumentsDesc?: string[]
}

export interface MethodResult {
  name: string
  describe?: string[]
  argumentsDesc?: string[]
}

export type AttrsMap = {
  [key: string]: string
}

export interface SlotResult {
  name: string
  describe: string
  backerDesc: string
  bindings: AttrsMap
}

export interface ParserOptions {
  onProp?: {
    (propsRes?: PropsResult[]): any
  }
  onEvent?: {
    (eventRes?: EventResult): any
  }
  onMethod?: {
    (methodRes?: MethodResult): any
  }
  onSlot?: {
    (slotRes?: SlotResult): any
  }
  onName?: {
    (name: string): any
  }
  babelParserPlugins?: ParserPlugin[]
}

export interface ParserResult {
  props?: PropsResult[]
  events?: EventResult[]
  slots?: SlotResult[]
  methods?: MethodResult[]
  name?: string
}

export default function(
  source: string,
  options: ParserOptions = {}
): ParserResult {
  const astRes = sfcToAST(source, options.babelParserPlugins)
  const res: ParserResult = {}
  const defaultOptions: ParserOptions = {
    onName(name: string) {
      res.name = name
    },
    onProp(propsRes?: PropsResult[]) {
      if (propsRes) {
        res.props = propsRes
      }
    },
    onEvent(eventsRes?: EventResult) {
      if (eventsRes) {
        ;(res.events || (res.events = [])).push(eventsRes)
      }
    },
    onSlot(slotRes?: SlotResult) {
      if (slotRes) {
        ;(res.slots || (res.slots = [])).push(slotRes)
      }
    },
    onMethod(methodRes?: MethodResult) {
      if (methodRes) {
        ;(res.methods || (res.methods = [])).push(methodRes)
      }
    },
    babelParserPlugins: ['objectRestSpread', 'dynamicImport']
  }

  const finallyOptions: ParserOptions = Object.assign(defaultOptions, options)
  if (astRes.jsAst) {
    parseJavascript(astRes.jsAst, finallyOptions)
  }
  if (astRes.templateAst) {
    parseTemplate(astRes.templateAst, finallyOptions)
  }

  return res
}
