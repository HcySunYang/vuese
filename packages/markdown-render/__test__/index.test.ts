import Render, { RenderResult, MarkdownResult } from '@vuese/markdown-render'
import { ParserResult } from '@vuese/parser'

test('Proper rendering of the table header', () => {
  const res: ParserResult = {
    name: 'MyComponent',
    componentDesc: {
      default: ['This is a description of the component'],
      group: ['My Group']
    },
    props: [
      {
        name: 'someProp',
        type: ['String'],
        typeDesc: ['`TOP` / `BOTTOM`'],
        required: true,
        defaultDesc: ['`TOP`'],
        describe: ['Represents the direction of the arrow']
      }
    ],
    events: [
      {
        name: 'click',
        isSync: false,
        syncProp: '',
        describe: ['Triggered when clicked'],
        argumentsDesc: ['a boolean value']
      }
    ],
    slots: [
      {
        name: 'header',
        describe: 'Table header',
        backerDesc: '`<th>{{title}}</th>`',
        bindings: {},
        scoped: false,
        target: 'template'
      },
      {
        name: 'header',
        describe: 'Table header',
        backerDesc: '`<th>{{title}}</th>`',
        bindings: {},
        scoped: false,
        target: 'script'
      }
    ],
    methods: [
      {
        name: 'clear',
        describe: ['Clear form'],
        argumentsDesc: ['a boolean value']
      }
    ],
    mixIns: [
      {
        mixIn: 'MixIn'
      }
    ],
    data: [
      {
        name: 'someVariable',
        type: 'String',
        describe: ['Does something'],
        default: ''
      }
    ],
    watch: [
      {
        name: 'question',
        describe: ['Something about a question'],
        argumentsDesc: ['a string value']
      }
    ]
  }
  const render = new Render(res)
  const renderRes: RenderResult = render.render()
  const markdownRes = render.renderMarkdown() as MarkdownResult
  expect(renderRes).toMatchSnapshot()
  expect(markdownRes).toMatchSnapshot()
})

test('Proper rendering of store', () => {
  const res: ParserResult = {
    name: 'Store',
    getters: [
      {
        name: 'formattedObject',
        describe: ['Returns a correctly formatted object']
      }
    ],
    actions: [
      {
        name: 'fetchObject',
        describe: ['Grabs the action from the API'],
        argumentsDesc: ['the url to get']
      }
    ],
    mutations: [
      {
        name: 'MUTATE_OBJECT',
        describe: ['Set the object in the store'],
        argumentsDesc: ['object with key/value to set']
      }
    ],
    state: [
      {
        name: 'stateObject',
        describe: ['stores the object']
      }
    ]
  }
  const render = new Render(res)
  const renderRes: RenderResult = render.render()
  const markdownRes = render.renderMarkdown() as MarkdownResult
  expect(renderRes).toMatchSnapshot()
  expect(markdownRes).toMatchSnapshot()
})

describe('Empty Components (without slots,methods,props,events) can be forced to be rendered', () => {
  test('Does not render an empty component by defaut', () => {
    const res: ParserResult = {
      name: 'MyComponent',
      componentDesc: {
        default: ['This is a description of the component']
      }
    }
    const render = new Render(res)
    const renderRes: RenderResult = render.render()
    const markdownRes = render.renderMarkdown() as MarkdownResult
    expect(markdownRes).toBeNull()
    expect(renderRes).toEqual({})
  })
  test('Does render an empty component by with @vuese in the description', () => {
    const res: ParserResult = {
      name: 'MyComponent',
      componentDesc: {
        default: [
          'This is a description of the component.',
          'Supplementary description'
        ],
        vuese: ['']
      }
    }
    const render = new Render(res)
    const renderRes: RenderResult = render.render()
    const markdownRes = render.renderMarkdown() as MarkdownResult
    expect(markdownRes).not.toBeNull()
    expect(markdownRes).toMatchSnapshot()
    expect(renderRes).toMatchSnapshot()
  })
})
