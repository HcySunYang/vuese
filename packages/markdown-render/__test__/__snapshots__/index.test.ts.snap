// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty Components (without slots,methods,props,events) can be forced to be rendered Does render an empty component by with @vuese in the description 1`] = `
Object {
  "componentName": "MyComponent",
  "content": "# MyComponent

This is a description of the component. Supplementary description

",
  "groupName": "BASIC",
}
`;

exports[`Empty Components (without slots,methods,props,events) can be forced to be rendered Does render an empty component by with @vuese in the description 2`] = `Object {}`;

exports[`Proper rendering of store 1`] = `
Object {
  "actions": "|Action|Description|Parameters|
|---|---|---|
|fetchObject|Grabs the action from the API|the url to get|
",
  "getters": "|Getter|Description|
|---|---|
|formattedObject|Returns a correctly formatted object|
",
  "mutations": "|Mutation|Description|Parameters|
|---|---|---|
|MUTATE_OBJECT|Set the object in the store|object with key/value to set|
",
  "state": "|Name|Description|
|---|---|
|stateObject|stores the object|
",
}
`;

exports[`Proper rendering of store 2`] = `
Object {
  "componentName": "Store",
  "content": "# Store

## Getters

<!-- @vuese:Store:getters:start -->
|Getter|Description|
|---|---|
|formattedObject|Returns a correctly formatted object|

<!-- @vuese:Store:getters:end -->


## Actions

<!-- @vuese:Store:actions:start -->
|Action|Description|Parameters|
|---|---|---|
|fetchObject|Grabs the action from the API|the url to get|

<!-- @vuese:Store:actions:end -->


## Mutations

<!-- @vuese:Store:mutations:start -->
|Mutation|Description|Parameters|
|---|---|---|
|MUTATE_OBJECT|Set the object in the store|object with key/value to set|

<!-- @vuese:Store:mutations:end -->


## State

<!-- @vuese:Store:state:start -->
|Name|Description|
|---|---|
|stateObject|stores the object|

<!-- @vuese:Store:state:end -->


",
  "groupName": "BASIC",
}
`;

exports[`Proper rendering of the table header 1`] = `
Object {
  "data": "|Name|Type|Description|Default|
|---|---|---|---|
|someVariable|\`String\`|Does something|-|
",
  "events": "|Event Name|Description|Parameters|
|---|---|---|
|click|Triggered when clicked|a boolean value|
",
  "methods": "|Method|Description|Parameters|
|---|---|---|
|clear|Clear form|a boolean value|
",
  "mixIns": "|MixIn|
|---|
|MixIn|
",
  "props": "|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|someProp|Represents the direction of the arrow|\`TOP\` / \`BOTTOM\`|\`true\`|\`TOP\`|
",
  "slots": "|Name|Description|Default Slot Content|
|---|---|---|
|header|Table header|\`<th>{{title}}</th>\`|
",
  "watch": "|Name|Description|Parameters|
|---|---|---|
|question|Something about a question|a string value|
",
}
`;

exports[`Proper rendering of the table header 2`] = `
Object {
  "componentName": "MyComponent",
  "content": "# MyComponent

This is a description of the component

## Props

<!-- @vuese:MyComponent:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|someProp|Represents the direction of the arrow|\`TOP\` / \`BOTTOM\`|\`true\`|\`TOP\`|

<!-- @vuese:MyComponent:props:end -->


## Events

<!-- @vuese:MyComponent:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|click|Triggered when clicked|a boolean value|

<!-- @vuese:MyComponent:events:end -->


## Slots

<!-- @vuese:MyComponent:slots:start -->
|Name|Description|Default Slot Content|
|---|---|---|
|header|Table header|\`<th>{{title}}</th>\`|

<!-- @vuese:MyComponent:slots:end -->


## Methods

<!-- @vuese:MyComponent:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|clear|Clear form|a boolean value|

<!-- @vuese:MyComponent:methods:end -->


## MixIns

<!-- @vuese:MyComponent:mixIns:start -->
|MixIn|
|---|
|MixIn|

<!-- @vuese:MyComponent:mixIns:end -->


## Data

<!-- @vuese:MyComponent:data:start -->
|Name|Type|Description|Default|
|---|---|---|---|
|someVariable|\`String\`|Does something|-|

<!-- @vuese:MyComponent:data:end -->


## Watch

<!-- @vuese:MyComponent:watch:start -->
|Name|Description|Parameters|
|---|---|---|
|question|Something about a question|a string value|

<!-- @vuese:MyComponent:watch:end -->


",
  "groupName": "My Group",
}
`;
